File: js-template/leetcode/editor/cn/subarray-sum-equals-k.js
/*
 * @lc app=leetcode.cn id=560 lang=javascript
 * @lcpr version=30404
 *
 * [560] 和为 K 的子数组
 */


import {ListNode} from "../common/listNode.js";
import {TreeNode} from "../common/treeNode.js";

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    var left = 0;
    var right = 0;
    var result = 0;

    if (nums.length === 1) {
        if (nums[0] === k)
            return 1;
        else
            return 0;
    }

    for (let i = 0; i < nums.length; i++) {
        const head = nums[i];
        let sum = head;
        if (sum === k)
            result++;
        for (let j = i + 1; j < nums.length; j++) {
            const element = nums[j];
            sum += element;
            if (sum === k)
                result++;
        }
    }

    return result;
};
// @lc code=end

// your test code here

console.log(subarraySum([-1, -1, 1], 0))



/*
// @lcpr case=start
// [1,1,1]\n2\n
// @lcpr case=end

// @lcpr case=start
// [1,2,3]\n3\n
// @lcpr case=end

 */

File: cpp-template/leetcode/editor/cn/subarray-sum-equals-k.cpp
/*
 * @lc app=leetcode.cn id=560 lang=cpp
 * @lcpr version=30404
 *
 * [560] 和为 K 的子数组
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

using namespace std;

// @lc code=start
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_multiset<int> cache;
        vector<int> sums;
        int sum = 0, result = 0;
        cache.insert(0);
        for (auto &&num : nums) {
            sum += num;

            int target = sum - k;
            result += cache.count(target);

            sums.push_back(sum);
            cache.insert(sum);
        }
        return result;
    }
};
// @lc code=end

int main() {
    Solution solution;
    // your test code here
    vector<int> nums = {-1, -1, 1};
    // vector<int> nums = {1,2,3};
    solution.subarraySum(nums, 0);
}



/*
// @lcpr case=start
// [1,1,1]\n2\n
// @lcpr case=end

// @lcpr case=start
// [1,2,3]\n3\n
// @lcpr case=end

 */

